### PR TITLE
fix(aws): upgrade to use non-deprecated attributes

### PR DESCRIPTION
**Pull request by qbiqing**
There was a recent change on removing deprecated attributes from the AWS 
provider (see https://github.com/terraform-providers/terraform-provider-aws/issues/13398), 
because the GitHub CI uses aws-provider v3.0.0, the builds would not pass 
after the change.

### CODE REVIEW

File: modules/core/elb.tf [hunk 1]
```diff
@@ -110,8 +110,9 @@ resource "aws_lb_listener_rule" "nomad_server" {
   }
 
   condition {
-    field  = "host-header"
-    values = [var.nomad_api_domain]
+    host-header {
+      values = [var.nomad_api_domain]
+    }
```
Comment (qbiqing) [inline modules/core/elb.tf:115]: For this change, see: 
Documentation for new syntax: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener_rule
Issue and when it was merged: 
https://github.com/terraform-providers/terraform-provider-aws/issues/8126

File: modules/lambda-api-gateway/api-gateway.tf [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
+
+resource "aws_api_gateway_usage_plan" "ApiKey" {
+  name = var.api_key_name
 
-  stage_key {
-    rest_api_id = aws_api_gateway_rest_api.api-gateway.id
-    stage_name  = aws_api_gateway_deployment.api-gateway-deployment.stage_name
+  api_stages {
+    api_id = aws_api_gateway_rest_api.api-gateway.id
+    stage  = aws_api_gateway_deployment.api-gateway-deployment.stage_name
```
Comment (qbiqing) [inline modules/lambda-api-gateway/api-gateway.tf:10]: Documentation on switching to new syntax: https://www.devopsschool.com/pdf/terraform/slides/Complete-Terraform-documentation/aws.pdf

**Comment by guangie88** [comment]
Can you help to add a provider versioning constraint on `aws`? Like this:

```tf
terraform {
  required_providers {
    aws = ">= 2.42"
  }
}
```

I chose the above version number based on first support for the new syntax, based on https://github.com/terraform-providers/terraform-provider-aws/issues/8126. Let me know if it's not correct?

Ref: https://www.terraform.io/docs/configuration/providers.html#provider-versions

**Comment by qbiqing** [comment]
> Can you help to add a provider versioning constraint on `aws`? Like this:
> 
> ```terraform
> terraform {
>   required_providers {
>     aws = ">= 2.42"
>   }
> }
> ```
> 
> I chose the above version number based on first support for the new syntax, based on [terraform-providers/terraform-provider-aws#8126](https://github.com/terraform-providers/terraform-provider-aws/issues/8126). Let me know if it's not correct?
> 
> Ref: https://www.terraform.io/docs/configuration/providers.html#provider-versions

Did a search through the codebase and updated all `aws` provider versions.

**Pull request merged**